Standardise development scripts

djbingham opened the pull request:
https://eaflood.atlassian.net/browse/FPD-284

All FFC microservices should have standardised scripts for common development tasks, such as setup, start, stop and test. This will enable a product development repository to reliably execute commands against all the repositories making up that product, such as to start a local instance of the entire product stack.

Commit:
Remove restriction on which services to build/run

https://eaflood.atlassian.net/browse/FPD-284

This is mainly for consistency with other repositories where more than
one service is started from a single docker-compose command. Restricting
docker-compose commands to specified services is likely to trip up
developers if supporting services (e.g. a data store) are added to this
repository in future.

Commit:
Consolidate scripts using docker-compose

- Remove scripts/start-compose as it duplicates scripts/start.
- Rename scripts/test-compose to remove `-compose` since this should be
the only test script.

Commit:
Improve readme formatting

Commit:
Standardise scripts

Commit:
Correct service name in PR testing instructions

Commit:
Run tests through docker-compose

This makes the test script cleaner as the volume mounts are taken care
of in docker-compose.yaml.

Commit:
Standardise scripts in package.json

Commit:
Ignore .config directory created by NPM

Commit:
Use /home/node as project root in Docker image

Commit:
Ensure node user owns all files in Docker image

Commit:
Merge remote-tracking branch 'origin/master' into full-stack-development

Rebase proved much too complicated due to many conflicts in specific
commits which did not appear in the full merge.

In the full-stack-development branch, docker-compose.test.yaml proved to
be redundant once the scripts were streamlined and standardised. Tests
are now run via npm in the main container image.

Commit:
Stop bind-mounting project root to container in CI

Commit:
Save test output to host form CI tests

Commit:
Add note on purpose of docker-compose.local.yml

Commit:
Ignore .standard-v12-cache folder generated by NPM

Commit:
Make scripted docker-compose commands more clear

Commit:
Use test override Docker Compose file in CI test

markysoft commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
-WORKDIR /usr/src/app
+USER node
+WORKDIR /home/node
 
```
we agreed that the standard directoty should be `/user/src/app`. Looks like this is missing from our standards, so I've added it at https://eaflood.atlassian.net/wiki/spaces/FFCPD/pages/1297219944/Standards+for+Development+Project+Structure+and+Naming+Conventions

djbingham replied:
I must have missed that discussion. I noticed that the repositories weren't using a consistent location and while investigating a permission issue I noticed that the working directories they used weren't owned by the user the containers run as. I changed them all on the basis of `/home/node` being owned by the `node` user so it reduces the chance of permission errors without needing additional steps in the Dockerfile to change the ownership.

I'm happy to change it to the agreed standard but I am curious - what was the reasoning for `/usr/src/app`?

johnwatson484 replied:
It was a discussion with myself, Mark and Ant.  Ant raised concerned that we weren't using /usr/src/app as it was generally considered the standard (albeit a still debated standard), he linked an article which has since disappeared from Slack.  There also seem to be some open source tools that expect that to be the location.

As we have no strong reason not to use this location, we agreed we would.

djbingham replied:
I've now reverted back to `/usr/src/app` and added the extra line to make `node` the owner of that folder.

markysoft commented on package.json:
```diff
@@ -4,12 +4,11 @@
+    "test": "npm run test:unit",
+    "test:lint": "standard",
+    "test:unit": "lab -c -r console -o stdout -r junit -o ./test-output/junit.xml",
+    "watch:server": "nodemon --inspect=0.0.0.0 --ext js --legacy-watch index.js"
```
Have we agreed on a standard of having a ':' in the script name?
I haven't got anything against the format, but want to make sure we are all on the same page.
Seems like some Defra repos use a colon, while others do not.

djbingham replied:
This was agreed within the ELM team during our previous prototype development but apparently not with the wider FFC team. Discussed via Slack, it seems everyone is happy to stick with the colon format.

Commit:
Use atomic Helm deployments

Prevents issues after failed Helm deploy.

markysoft commented on azure-pipelines.yaml:
```diff
@@ -57,40 +57,30 @@ steps:
     projectName: $(Build.Repository.Name)
     qualifyImageNames: true
-    dockerComposeCommand: -f docker-compose.test.yaml build --no-cache $(imageName)
-  displayName: build tests image
```
I would prefer this step to be left in.
There should be a difference between the test image and the base image as the final image will not include dev dependencies.
Rather than remove the step it might be better to change the `build base image` step to build a production image with no dev dependencies.
If we take the step out we'll have to put it back in again when the test and base images diverge.

djbingham replied:
I agree the production build should not include dev dependencies. We have a separate ticket in the backlog to address the Dockerfiles. Whether we need two build stages in the pipeline depends on how we end up structuring the Dockerfile and docker-compose configurations.

For now, I felt that having two stages build the same image was unnecessarily slowing down the pipeline.

djbingham replied:
On the ELM prototype we didn't end up needing two build stages in the pipeline, even after stripping dev tools out of the production build. I think this was because we ran a production-like app container and a separate "watcher" container in development so docker-compose built both images from a single command.

Looking back I think this approach relied on our use of Skaffold to restart the app container on code changes. As we've moved away from Skaffold to docker-compose, we may well end up needing two build stages in the pipeline now.

djbingham replied:
I've restored the separate build steps for test and base images for now. We do need to think more about this when the Dockerfile gets improved (there's already a ticket for that in the backlog). Building the production image from scratch after running tests against a different image carries a risk of divergence so that tests lose value over time. Ideally the production image should be built first with the test build extending from it.

markysoft replied:
The package-lock.json specifies a version, location and integrity hash for every module and each of its dependencies. The install it creates will be the same, every single time.
If the npm install uses package-lock there will be no divergence.

markysoft commented on azure-pipelines.yaml:
```diff
@@ -57,40 +57,30 @@ steps:
     azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
     azureContainerRegistry: $(azureContainerRegistryFull)
     dockerComposeFile: docker-compose.yaml
+    additionalDockerComposeFiles: docker-compose.test.yaml
```
good spot on the additional config options in the DevOps tasks

markysoft commented on azure-pipelines.yaml:
```diff
@@ -57,40 +57,30 @@ steps:
     projectName: $(Build.Repository.Name)
     qualifyImageNames: true
-    dockerComposeCommand: -f docker-compose.test.yaml up --force-recreate  $(imageName)
+    dockerComposeCommand: run $(imageName) npm run test
```
rather than using run and specifing the command here it could be set in the `docker-compose.test.yaml`

djbingham replied:
Fair point. This just came from iterating through other solutions before arriving back at the separate test compose file.

djbingham replied:
I've changed this to `dockerComposeCommand: run $(imageName)` and set the command to `npm run test` in the Docker Compose file.

Matthew-Collins-Defra commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
 FROM node:10.15.3-alpine
 
-# Create app directory
-WORKDIR /usr/src/app
```
This was changed in a previous PR, just wondering why it's changing again? There was some discussion about standardising on /usr/src/app previously.
https://github.com/DEFRA/mine-support/commit/86fefec4073beaaa99b6547e2b839bfaa73eb925

djbingham replied:
I wasn't aware of that. I switched to a folder owned by the `node` user after dealing with a permission issue in the build pipeline but we can make it work with `/usr/src/app` if that's what we want to use.

markysoft commented on docker-compose.test.yaml:
```diff
@@ -1,6 +1,9 @@
-  mine-support-api-gateway:
-    volumes:
-      - ./test-output:/usr/src/app/test-output
-    command: npm run test
```
why remove the command? Doing so means it has to be set explicitly in the test script and the pipeline step when the `docker-compose.test.yaml`

djbingham replied:
Good point. I went through a series of iterations where it made sense to remove the test compose file altogether but eventually came round to adding it back in. I find it quite odd to use `up` rather than `run` for this purpose but putting the npm command in one place is a good benefit.

djbingham replied:
It turns out `run` doesn't require a command to be specified anyway, it falls back to the one in the compose file or the Dockerfile. I've put the `npm run` command back into `docker-compose.test.yml`.

markysoft commented on helm/values.yaml:
```diff
@@ -16,7 +16,7 @@ container:
   runAsUser: 1000
   readOnlyRootFilesystem: true
   allowPrivilegeEscalation: false
-  userServiceUrl: http://mine-support-user-service.default
```
The default userServiceUrl has been reverted to use port 3002. The user service was changed to run on port 80, so the default URL should be  `userServiceUrl: http://mine-support-user-service.default`

djbingham replied:
When/where was the user service updated? It still defaults to 3002 in the [config.js](https://github.com/DEFRA/mine-support-user-service/blob/master/server/config.js)

johnwatson484 replied:
For Helm deployments the services all use port 80, when running locally they use the 300x values

djbingham replied:
Checked with @johnwatson484, the update was specifically to Helm deployments.

I've reverted back to the correct port.

markysoft reviewed (changes requested):
There are a number of changes to script names in the package.json that I would like to confirm that we have agreed on.

The user service URL in the vaules file has been reverted, this should change back to being on port 80.

The build test image step has been removed from the pipeline. 
Rather than removing a step we would need if the builds were set up correctly, it would be better to change the build base image step to correctly build a production image without dev dependencies.

Commit:
Specify test command in docker-compose.test.yml

Commit:
Restore the test build pipeline step

This will be required when the Dockerfile is refactored to produce
production builds without dev dependencies. Mark H suggested it would be
better to keep the stage in rather than remove it now and add it back
later.

Commit:
Set Docker image working directory to /usr/src/app

This has been agreed as the standard location for app files in Docker.

Commit:
Remove additional tag from test Docker image in CI

This was erroneously included in a copy+paste.

Commit:
Revert user service URL to use port 80 in Helm

Commit:
Rename NPM script watch:server to start:watch

Commit:
Return CI base image build to after test run

This ensures the test build doesn't override the base image.
Needs revisiting when the Dockerfile is refactored as the test build
should extend the base image without overwriting it.

johnwatson484 reviewed (approved):
Confirmed build has passed and can run and test locally.
